### PR TITLE
Implement hardcoded sync in the light client

### PR DESCRIPTION
**Pull request by tomaka**
Adds a `hardcodedSync` field to the chain specifications.
When set, contains the RLP of a header, a total difficulty, and a list of CHT root hashes. If there is no light client database, the light client will immediately jump to the given block.

After this change, the light client synchronizes in 10 to 15 seconds even at the first launch on the main network.

For now only the specs of the main network have been updated.

### CODE REVIEW

File: ethcore/light/src/client/header_chain.rs [hunk 1]
```diff
@@ -270,15 +271,47 @@ impl HeaderChain {
...
+				}
+
+				let decoded_header = hardcoded_sync.header.decode();
+
+				// check that we have enough hardcoded CHT roots. avoids panicking later.
+				{
+					let cht_num = cht::block_to_cht_number(decoded_header.number() - 1)
+						.expect("specs provided a hardcoded block with height 0");
+					if cht_num >= hardcoded_sync.chts.len() as u64 {
+						panic!("specs didn't provide enough CHT roots for its hardcoded block");
```
Comment (rphmeier) [inline ethcore/light/src/client/header_chain.rs]: rule for panics of any kind is "prove or remove" -- We may rather want to print a warning and fall back to hardcoded-less sync. Or move these checks to the deserialization phase where we can exit with an error more gracefully.

File: ethcore/light/src/client/header_chain.rs [hunk 2]
```diff
@@ -299,11 +332,36 @@ impl HeaderChain {
 	///
 	/// If the block is an epoch transition, provide the transition along with
 	/// the header.
+	#[inline]
```
Comment (rphmeier) [inline ethcore/light/src/client/header_chain.rs]: does that actually help? generally I figure LLVM will figure it out.

File: ethcore/light/src/client/header_chain.rs [hunk 3]
```diff
@@ -479,6 +542,49 @@ impl HeaderChain {
...
+
+		for cht_num in 0.. {
+			let cht = match self.cht_root(cht_num) {
+				Some(cht) => cht,
+				_ if cht_num != 0 => {
+					let h_num = 1 + cht_num as u64 * cht::SIZE;
+					let header = if let Some(header) = self.block_header(BlockId::Number(h_num)) {
+						header
+					} else {
+						panic!("Header of block #{} not found in DB", h_num);
```
Comment (rphmeier) [inline ethcore/light/src/client/header_chain.rs]: same: do not panic.

File: ethcore/light/src/client/header_chain.rs [hunk 4]
```diff
@@ -479,6 +542,49 @@ impl HeaderChain {
...
+					};
+
+					let decoded = header.decode();
+
+					let entry: Entry = ::rlp::decode(&self.db.get(self.col, era_key(h_num).as_bytes()).unwrap().unwrap());
```
Comment (rphmeier) [inline ethcore/light/src/client/header_chain.rs]: and we really never use unwrap.

File: ethcore/light/src/client/header_chain.rs [hunk 5]
```diff
@@ -479,6 +542,49 @@ impl HeaderChain {
...
+					let total_difficulty = entry.candidates.iter()
+						.find(|c| c.hash == decoded.hash())
+						.expect("no candidate matching block found in DB")
```
Comment (rphmeier) [inline ethcore/light/src/client/header_chain.rs]: should rather prove based on the fact that an inconsistent but correctly-loaded database shouldn't be possible.

File: ethcore/light/src/client/header_chain.rs [hunk 6]
```diff
@@ -479,6 +542,49 @@ impl HeaderChain {
...
+				},
+				_ => {
+					return None;
+				},
+			};
+
+			chts.push(cht);
+		}
+
+		unreachable!("we are after an infinite loop")
```
Comment (rphmeier) [inline ethcore/light/src/client/header_chain.rs]: could be a `loop` with a break-value

File: parity/run.rs [hunk 7]
```diff
@@ -258,6 +258,11 @@ fn execute_light_impl(cmd: RunCmd, can_restart: bool, logger: Arc<RotatingLogger
 	let service = light_client::Service::start(config, &spec, fetch, db, cache.clone())
 		.map_err(|e| format!("Error starting light client: {}", e))?;
 	let client = service.client();
+	// Note: uncomment this code if you want to regenerate the JSON of the hardcoded sync feature
+	// in the chain specifications.
+	/*if let Some(hs) = client.read_hardcoded_sync() {
```
Comment (rphmeier) [inline parity/run.rs]: we could make this a subcommand of some kind

**Comment by rphmeier** [comment]
needs an opt-out flag also (off by default is fine, just that it should be there).

File: parity/run.rs (hunk 7 above)
Comment (rphmeier) [reply parity/run.rs]: generally not a fan of the commented-out code. the CLI is a PITA to work with though :\

**Comment by rphmeier** [comment]
looks good other than grumbles

**Comment by tomaka** [comment]
Should I make `read_hardcoded_sync()` return `None` in case of error? Or rather `Result<Option<...>, ...>`? I admit that I didn't really know what to do with it, it's an internal-ish method.

**Comment by rphmeier** [comment]
`Result<Option<_>>` would be better IMO

**Comment by tomaka** [comment]
Sorry for the bad PR quality. I think it should be good now!

**Comment by rphmeier** [comment]
needs merge

**Comment by tomaka** [comment]
@rphmeier Conflicts solved!

File: parity/export_hardcoded_sync.rs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+		no_hardcoded_sync: true,
+	};
+
+	config.queue.max_mem_use = cmd.cache_config.queue() as usize * 1024 * 1024;
+
+	// start on_demand service.
+	let on_demand = Arc::new(::light::on_demand::OnDemand::new(cache.clone()));
+
+	let sync_handle = Arc::new(RwLock::new(Weak::new()));
+	let fetch = ::light_helpers::EpochFetch {
```
Comment (rphmeier) [inline parity/export_hardcoded_sync.rs]: on_demand, sync, network etc. are not necessary to start here -- a `NoOpEpochFetch` (I think it already exists for tests) will be enough.

**Comment by rphmeier** [comment]
LGTM except for grumble

**Comment by tomaka** [comment]
r?

File: ethcore/res/ethereum/foundation.json [hunk 9]
```diff
@@ -173,6 +173,2593 @@
...
+		"totalDifficulty": "3144406426008470895785",
+		"CHTs": [
+			"0x0eb474b7721727204978e92e27d31cddff56471911e424a4c8271c35f9c982cc",
```
Comment (5chdn) [inline ethcore/res/ethereum/foundation.json]: Is this anyhow possible without adding 2500 lines to the chain spec?

Comment (tomaka) [reply ethcore/res/ethereum/foundation.json]: Hmmm, what would the alternative be? We're hardcoding headers in the client after all, we have to put them *somewhere*.

Comment (rphmeier) [reply ethcore/res/ethereum/foundation.json]: the alternative is to allow the client to be deployed w/o history or CHT roots and have the history fetched afterwards

**Pull request merged**